File: app/src/main/java/com/imangazalievm/bubbble/presentation/mvp/presenters/UserProfilePresenter.java
package com.imangazalievm.bubbble.presentation.mvp.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.imangazalievm.bubbble.domain.exceptions.NoNetworkException;
import com.imangazalievm.bubbble.domain.interactors.UserProfileInteractor;
import com.imangazalievm.bubbble.domain.models.User;
import com.imangazalievm.bubbble.presentation.commons.rx.RxSchedulersProvider;
import com.imangazalievm.bubbble.presentation.mvp.views.UserProfileView;
import com.imangazalievm.bubbble.presentation.utils.DebugUtils;

import javax.inject.Inject;

@InjectViewState
public class UserProfilePresenter extends MvpPresenter<UserProfileView> {


    private UserProfileInteractor userProfileInteractor;
    private RxSchedulersProvider rxSchedulersProvider;
    private long userId;
    private User user;

    @Inject
    public UserProfilePresenter(UserProfileInteractor userProfileInteractor, RxSchedulersProvider rxSchedulersProvider, long userId) {
        this.userProfileInteractor = userProfileInteractor;
        this.rxSchedulersProvider = rxSchedulersProvider;
        this.userId = userId;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        loadUser();
    }

    private void loadUser() {
        getViewState().showLoadingProgress();
        userProfileInteractor.getUser(userId)
                .compose(rxSchedulersProvider.getIoToMainTransformerSingle())
                .subscribe(this::onUserLoaded, this::onUserLoadError);
    }

    private void onUserLoaded(User user) {
        this.user = user;
        getViewState().hideLoadingProgress();
        getViewState().showUser(user);
    }

    private void onUserLoadError(Throwable throwable) {
        if (throwable instanceof NoNetworkException) {
            getViewState().hideLoadingProgress();
            getViewState().showNoNetworkLayout();
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

    public void onShareProfileClicked() {
        getViewState().showUserProfileSharing(user);
    }

    public void onOpenInBrowserClicked() {
        getViewState().openInBrowser(user.getHtmlUrl());
    }

    public void onLinkClicked(String url) {
        getViewState().openInBrowser(url);
    }

    public void onUserSelected(long userId) {
        getViewState().openUserProfileScreen(userId);
    }

    public void retryLoading() {
        getViewState().hideNoNetworkLayout();
        getViewState().showLoadingProgress();
        loadUser();
    }

    public void onBackPressed() {
        getViewState().closeScreen();
    }

}

File: app/src/main/java/com/imangazalievm/bubbble/presentation/userprofile/UserProfilePresenter.kt
package com.imangazalievm.bubbble.presentation.userprofile

import com.arellomobile.mvp.InjectViewState
import com.imangazalievm.bubbble.data.global.network.exceptions.NoNetworkException
import com.imangazalievm.bubbble.domain.global.models.User
import com.imangazalievm.bubbble.domain.userprofile.UserProfileInteractor
import com.imangazalievm.bubbble.presentation.global.mvp.BasePresenter
import javax.inject.Inject

@InjectViewState
class UserProfilePresenter @Inject constructor(
    private val userProfileInteractor: UserProfileInteractor,
    private val userId: Long
) : BasePresenter<UserProfileView>() {

    private lateinit var user: User
    private val isUserLoaded: Boolean
        get() = ::user.isInitialized

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadUser()
    }

    private fun loadUser() = launchSafe {
        viewState.showLoadingProgress()
        try {
            this@UserProfilePresenter.user = userProfileInteractor.getUser(userId)
            viewState.showUser(user)
        } catch (throwable: NoNetworkException) {
            viewState.showNoNetworkLayout()
        } finally {
            viewState.hideLoadingProgress()
        }
    }

    fun retryLoading() {
        viewState.hideNoNetworkLayout()
        loadUser()
    }

    fun onShareProfileClicked() {
        if (!isUserLoaded) return
        viewState.showUserProfileSharing(user)
    }

    fun onOpenInBrowserClicked() {
        if (!isUserLoaded) return
        viewState.openInBrowser(user.htmlUrl)
    }

    fun onLinkClicked(url: String?) {
        viewState.openInBrowser(url!!)
    }

    fun onUserSelected(userId: Long) {
        viewState.openUserProfileScreen(userId)
    }

    override fun onBackPressed() {
        viewState.closeScreen()
    }

}